### PR TITLE
Use browser-polyfill.js to allow use with chromium

### PR DESCRIPTION
**Pull request by eNV25**
This change adds browser-polyfill.js so that it can also be packaged for chromium. This should result in no change in how the ff2mpv.js is executed on firefox.

More about browser-polyfill: https://github.com/mozilla/webextension-polyfill

- `./package-chromium.sh` generates dist.crx and dist.pem. The extension id is ~~generated from~~ tied to this private key file, keep it around so that it is reused and does not change.
- Load the extension in chromium by dragging and dropping it in the extension page.
- Make a ff2mpv-chromium.json manifest from ff2mpv-chromium.example.json. Replace the path with the correct path and `<your-extension-id>` with the extension id from `chrome://extensions`.
- Put the external manifest in the correct location: https://developer.chrome.com/docs/apps/nativeMessaging/#native-messaging-host-location
- You might need to enable developer mode in the extensions page. This should work in chromium builds in Linux distros, but I am not sure about official Google Chrome builds. You could also use "Load unpacked" to load this extension but this generates a new extension id every time so it's a pain in the ass.

If someone uploads this to the chrome webstore and keeps the private key file safe the extension id would stay the same, there would be no need for this dance and we could just have one ff2mpv-chrome.json manifest.

NOTE: This has nothing to do with the existing chrome port by @DanSM-5 https://github.com/DanSM-5/ff2mpv https://chrome.google.com/webstore/detail/ff2mpv/ephjcajbkgplkjmelpglennepbpmdpjg . In the future maybe he could upload this version of the chrome port to the chrome webstore, instead of using a fork.

### CODE REVIEW

File: package-zip.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
```
Comment (woodruffw) [inline package-zip.sh:1]: Mind renaming this to `package-firefox.sh`?

**Comment by woodruffw** [comment]
cc @DanSM-5: how do you feel about unifying on a single port of `ff2mpv` for Chrome?

File: ff2mpv-chromium.example.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+    "name": "ff2mpv",
+    "description": "ff2mpv's external manifest",
+    "path": "/home/william/scripts/ff2mpv",
+    "type": "stdio",
+    "allowed_origins": [
+        "chrome-extension://<your-extension-id>/"
```
[the remark this answers is not in the document]
Comment (eNV25) [reply ff2mpv-chromium.example.json:7]: `package-chromium.sh` generates a `.crx` extension file, and a `.pem` private key file. The private key file is reused when you run `package-chromium.sh` again and the extension id stays the same. If you delete the `.pem` private key file and run `package-chromium.sh` again a new `.pem` private key file is generated and the extension id changes. So AFAIK the extension id is tied to the private key file.

[the remark this answers is not in the document]
Comment (eNV25) [reply ff2mpv-chromium.example.json:7]: I am not sure how publishing to WebStore works but I know that extension-id usually stays the same.

**Comment by eNV25** [comment]
Something to note is Chrome's switch to manifest v3 and the manifest v2 support timeline. This extension uses manifest v2.

https://developer.chrome.com/docs/extensions/mv3/mv2-sunset/

image

It looks like after this month new mv2 extensions cannot be added to the Chrome Web Store, but old ones can be updated for a year. So the best bet for getting this in the Chrome Web Store in this time frame is updating @DanSM-5 fork with this version.

The long term strategy for supporting Chrome based browsers should be adopting manifest v3. But this will probably require some changes to the code (eg. how the extension finds the video from the current page). I will probably try to experiment with it at some point in the future.

AFAIK Firefox will keep supporting Manifest v2 and also support Manifest v3 in the future for Chrome compatibility. So even if the extension doesn't adopt manifest v3 it should still work with Firefox. Also Firefox nightly does not yet support Manifest v3.

**Comment by woodruffw** [comment]
Thanks for noting that. Yeah, in that case, unifying with the existing Chrome extension makes sense. That will require @DanSM-5 to either become the release manager for Chrome or for them to give me their current signing keys, in which case I don't mind taking over the responsibility.

And yeah, we should eventually switch to Manifest v3 overall. I think we can safely earmark that for the point in time when Firefox ESR supports it.

**Comment by DanSM-5** [comment]
I think it would be great to unify the development. I don't mind giving you the signing keys @woodruffw. I published it to the chrome store just to avoid updating the app id every time.

**Comment by woodruffw** [comment]
> I think it would be great to unify the development. I don't mind giving you the signing keys @woodruffw. I published it to the chrome store just to avoid updating the app id every time.

Awesome! We can arrange some kind of secure channel for the keys. Does Keybase or email + PGP work for you?

**Comment by DanSM-5** [comment]
I'm not familiar with keybase. Do you recommend it? I can encrypt it with PGP and send it over email.

**Comment by woodruffw** [comment]
Only if you don’t already have PGP set up! Email is good otherwise; my key ID is 0x85ae00c504833b3c.

**Comment by DanSM-5** [comment]
All set, let me know if you received it.

**Comment by woodruffw** [comment]
Received, thank you! I’ll try and decrypt this evening or tomorrow.

**Comment by eNV25** [comment]
Just to understand what is going, why is the private key being transferred? As far as I understand it would only allow you to locally sign the extension while having the same extension id as @DanSM-5 's fork. I don't think it allows you to publish to the chrome webstore.

https://stackoverflow.com/questions/5604491/how-do-i-transfer-the-ownership-of-a-chrome-extension

According to this Stack Overflow to transfer you need to use a form and have a valid destination account that has paid the $5 fee.

https://support.google.com/chrome_webstore/contact/dev_account_transfer

But it looks like you can also set up *group publishing* so multiple devs can manage one extension.

https://developer.chrome.com/docs/webstore/group-publishers/
https://developer.chrome.com/docs/webstore/group-publishers/#move-existing-items-to-a-group-publisher-account

Maybe that is also a good option?

**Comment by woodruffw** [comment]
> Just to understand what is going, why is the private key being transferred? As far as I understand it would only allow you to locally sign the extension while having the same extension id as @DanSM-5 's fork. I don't think it allows you to publish to the chrome webstore.

Oh, I badly misunderstood then -- I thought they key was needed for publishing to the webstore (I've never done any Chrome extension publishing/management before). Sorry for the confusion.

Group publishing looks like it's the right approach, although it's weirdly baked into Google Groups. I could create one (I have a G Suite account) if that works for you @DanSM-5, unless you'd prefer to transfer it outright.

**Comment by DanSM-5** [comment]
I was confused as well. I didn't know about the limitations of the Chrome extension store. Although I think the private key will be helpful if there is a plan to provide signed packages (crx) directly from the releases.

@woodruffw I agree with the group publishing. You can add me to a group and I will add the extension to it. Just want to mention this, it looks like there is a limit per account to create a single group publishing in the chrome store.

> A Chrome Web Store developer account can only create one group publisher, ever. You cannot replenish this quota, even if you delete the group.

If you don't want to use your quota on this, I can volunteer keep updating the extension in the chrome store once this PR gets merged.

**Comment by woodruffw** [comment]
Hooray for arbitrary restrictions! I probably won't ever write another extension, so I don't mind burning my quota on this. I'll set up the group and add you to it, so that there's an extra layer of redundancy for getting releases made for Chrome.

**Comment by woodruffw** [comment]
Okay, I've created the group and sent you an invite to it. Let me know if you have any problems with the invite.

**Comment by DanSM-5** [comment]
@woodruffw I waited all day but I didn't received any invitation to join. Could you try to add me again?

**Comment by woodruffw** [comment]
That's odd; it shows you as a member currently. I'll send you an email with a screenshot of what I'm seeing and we'll try to resolve over email (I don't want to disclose your email address or full name on GitHub without your permission, since they aren't immediately visible on your profile.)

**Comment by DanSM-5** [comment]
My bad, I am part of the group but I didn't get any notification about it. The group only appears if I try to make a transfer for the extension. I moved the extension to the group. You should have access to it now.

**Comment by woodruffw** [comment]
No problem; everything about this is Google's fault :wink: 

Yep, I also just confirmed that I have access to the extension. Now that that's all settled, I'll do another review pass of this PR and, with any luck, begin cutting release for both browsers this week!

**Comment by eNV25** [comment]
I will add a commit with a manifest json with the extension id already added.

File: ff2mpv-chromium.example.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply ff2mpv-chromium.example.json:1]: Nope, we can remove that.

**Review by woodruffw — APPROVED** [review]
LGTM! Once https://github.com/woodruffw/ff2mpv/pull/69/files#r790180598 is in, I'll merge this and cut a release for both platforms.

Thanks for all of the hard work here, and thanks again @DanSM-5 for the coordination.

**Comment by woodruffw** [comment]
Trying things out locally now. It looks like `snap`-distributed Chromium doesn't play nice with native messaging, so I need to install a non-`snap` version. But that's not our problem or something we want to solve (it's `snap`'s fault for having completely broken sandboxing).

**Comment by woodruffw** [comment]
Works locally for me! I'm going to merge, do a small amount of clean-up, and cut a new release for both platforms.

**Comment by woodruffw** [comment]
Cut as 4.0.0 and submitted for review to the Chrome Store. Will be published shortly on AMO as well.

**Pull request merged**